Userland/DevTools/UserspaceEmulator/Emulator.cpp

```diff
@@ -214,22 +216,75 @@ int Emulator::exec()
     constexpr bool trace = false;
 
     while (!m_shutdown) {
-        m_cpu.save_base_eip();
-
-        auto insn = X86::Instruction::from_stream(m_cpu, true, true);
+        if (m_steps_til_pause) {
+            // Exec cycle
+            m_cpu.save_base_eip();
+            auto insn = X86::Instruction::from_stream(m_cpu, true, true);
+            if constexpr (trace) {
+                outln("{:p}  \033[33;1m{}\033[0m", m_cpu.base_eip(), insn.to_string(m_cpu.base_eip(), symbol_provider));
+            }
 
-        if constexpr (trace) {
-            outln("{:p}  \033[33;1m{}\033[0m", m_cpu.base_eip(), insn.to_string(m_cpu.base_eip(), symbol_provider));
-        }
+            (m_cpu.*insn.handler())(insn);
 
-        (m_cpu.*insn.handler())(insn);
+            if constexpr (trace) {
+                m_cpu.dump();
+            }
 
-        if constexpr (trace) {
+            if (m_pending_signals) [[unlikely]] {
+                dispatch_one_pending_signal();
+            }
+            if (m_steps_til_pause > 0)
+                m_steps_til_pause--;
+
+        } else {
+            // Console interface
```

Comment:
Can you factor out the console stuff into a different function? It shouldn't just be sitting in the middle of `exec()` 😃

Reply:
its in `handle_REPL()` now, the name is a bit weird, but that should fix it for now